### ISSUE TITLE
Merge Item#applyMatrix & Item#matrix

### ISSUE THREAD

**Issue by lehni** [paperjs/paper.js#933]
This started off with an observation about JSON export, but turned into something bigger: I think we can get rid of the  `applyMatrix` values in the JSON export, simply by only including the matrix for items that have `applyMatrix == false`. If an item has `applyMatrix == true`, meaning it bakes its matrix into its content, the matrix is always the identity transformation, and hence does not need to be stored. The presence / absence of a stored matrix then directly reflects the setting for `applyMatrix`.

But then it occurred to me: We could then this a step further and say that if any item has its actual matrix set to `null`, it bakes any transformation into the content (if possible. if not, then it doesn't allow `null` for its matrix), and if it has the matrix set, the matrix will receive the transformation. For backward compatibility, `applyMatrix` could be kept around, and handle setting / erasing of `matrix`. The only change would be that `Item#matrix` could return `null`, while before it was always defined..

I was never a fan of `applyMatrix` which has caused a lot of confusion and keeps popping up on the mailing list (that's of course also a problem of documentation). But I feel that this proposal is much clearer and more intuitive as to what is actually happening.

The `paper.settings.applyMatrix` setting could be kept around, as it would define what happens to newly created items that are able to bake the transformations into their content (Group, Path, CompoundPath): Do they have their `matrix` set to `null` or the identity transformation when they are created?

Thoughts?

**Comment by bmacnaughton** [comment]
This makes sense to me. It will also, I think, result in more straightforward documentation because there is no need to explain the presence of a matrix when it is only the identity matrix.

I don't know how common a use case fetching the matrix is; I've never used it. If it is a common case then the getter could still return an identity matrix even though internally it is stored as null, right?

**Comment by lehni** [comment]
Yes but then you wouldn't have a way of checking if an item is using `applyMatrix` or not, kind of defeating the purpose I think?

There is one case I found where this proposal is currently problematic:

When I create an empty group, and apply some transformations to it, it stores them in the matrix even though `applyMatrix` is set to `true`, simply because it doesn't have any content to apply it to yet. When i then add the content, it bakes this stored matrix in. This behavior is even required by the SVG importer, I think. But I don't like it, as it's confusing and too much behind the scene magic. I think it would be good to have a clearer API behavior, I just need to find a better approach in the SVG importer for it...

**Comment by lehni** [comment]
Another thing to consider... If we want an item to not bake the transformations in, we'd have to explicitly set the matrix then: `item.matrix = new Matrix()`. That's also not so elegant...

**Comment by bmacnaughton** [comment]
As I haven't found myself using the transformation matrix much I haven't run into the situation of querying an item to find out if it is using a matrix or not. My only applications have been set/forget. So my comments all have that bias to them.

The empty group scenario being required by the SVG importer seems like an edge case. As you note, if you can find a way around that would be good. I don't think that it make sense for that to be the way things work in general, i.e., there's one special case when even though `applyMatrix` is `true` it doesn't really work that way.

The second issue - having to set the matrix to prevent transformations from being applied seems, as you say, not-so-elegant. If it is a rare use case then I think it's OK. But if it is a common use case then it seems like leaving things the way they are is no less elegant. My lack of experience using the transforms leaves me unable to offer an informed opinion.

On the one question you asked in the original post:

> The paper.settings.applyMatrix setting could be kept around, as it would define what happens to newly created items that are able to bake the transformations into their content (Group, Path, CompoundPath): Do they have their matrix set to null or the identity transformation when they are created?

I would opt for setting the matrix null, mostly as a point of consistency.
